core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala

```diff
@@ -59,13 +60,26 @@ private[spark] class ShuffleWriteProcessor extends Serializable with Logging {
         rdd.iterator(partition, context).asInstanceOf[Iterator[_ <: Product2[Any, Any]]])
       val mapStatus = writer.stop(success = true)
       if (mapStatus.isDefined) {
+        val isPushBasedShuffleEnabled = Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf,
+          isDriver = SparkContext.DRIVER_IDENTIFIER == SparkEnv.get.executorId)
+        // Check if sufficient shuffle mergers are available now for the ShuffleMapTask to push
+        if (isPushBasedShuffleEnabled && dep.getMergerLocs.isEmpty && !dep.shuffleMergeFinalized) {
```

Comment:
Review note: Here we are depending on the fact that retry of determinate stages will be merge finalized - while indeterminate stages wont be - and so the `!dep.shuffleMergeFinalized` takes care of not trying to fetch mergers/enable push based shuffle for retry of determinate stage.